services/brig/src/Brig/API/User.hs

```diff
@@ -253,6 +256,43 @@ createUserSpar new = do
       Team.TeamName nm <- lift $ liftSem $ GalleyAPIAccess.getTeamName tid
       pure $ CreateUserTeam tid nm
 
+upgradePersonalToTeam ::
+  forall r.
+  ( Member GalleyAPIAccess r,
+    Member UserSubsystem r,
+    Member TinyLog r,
+    Member (Embed HttpClientIO) r,
+    Member NotificationSubsystem r,
+    Member (Input (Local ())) r,
+    Member (Input UTCTime) r,
+    Member (ConnectionStore InternalPaging) r
+  ) =>
+  Local UserId ->
+  BindingNewTeamUser ->
+  ExceptT UpgradePersonalToTeamError (AppT r) CreateUserTeam
+upgradePersonalToTeam luid bNewTeam = do
+  -- check that the user is not part of a team
+  mSelfProfile <- lift $ liftSem $ getSelfProfile luid
+  let mTid = mSelfProfile >>= userTeam . selfUser
+  when (isJust mTid) $
+    throwE UpgradePersonalToTeamErrorAlreadyInATeam
+
+  lift $ do
+    -- generate team ID
+    tid <- randomId
+
+    let uid = tUnqualified luid
+    createUserTeam <- do
+      liftSem $ GalleyAPIAccess.createTeam uid (bnuTeam bNewTeam) tid
+      let BindingNewTeam newTeam = bNewTeam.bnuTeam
+      pure $ CreateUserTeam tid (fromRange (newTeam ^. newTeamName))
+
+    wrapClient $ updateUserTeam uid tid
+    liftSem $ Intra.sendUserEvent uid Nothing (teamUpdated uid tid)
+    initAccountFeatureConfig uid
+
+    pure $! createUserTeam
```

Comment:
Do we need to think about what can go wrong here if this fails in the middle of it and gets retried?
AFAICT the critical point is after the team ID of the user is updated because then it cannot be re-triggered. In that case, events might not get sent and `initAccountFeatureConfig` might not get called. How bad is this?

If it fails after team is created but before user team is updated, and then retried, a re-trigger should work AFAICT, there will just be an unused zombie team lying around in the DB.

Reply:
The same questions would apply to the code that creates a team owner from scratch, no? The order of operations is pretty much the same. As usual, in the absence of proper atomicity and transactions, I'm not sure we can ever come up with good answers to these questions.